README
stay tuned, coming soon..

CHANGELOG
openengsb-maven-plugin-2.3.1 2011-12-23
--------------------------------------------

This release was only required to make it possible to get a clean documentation on the homepage (which required
openengsb-root-28, which on the other hand required openengsb-maven-plugin-2.3.0).

### Details
** Library Upgrade
    * [OPENENGSB-2458] - Upgrade to root-28
    * [OPENENGSB-2459] - Reactivate source parsing

** Task
    * [OPENENGSB-2456] - Release openengsb-maven-plugin 2.3.1


openengsb-maven-plugin-2.3.0 2011-12-23
--------------------------------------------

The third feature release cut from the 2.x release series comes with great new features concerning the documentation.
With this plugin a new goal (extractSource) had been added which allows to tag source code and automatically generate
docbook src from it. In addition we've upgraded various libraries and included the documentation now directly into the
project. There's also a source package now providing the tagged source as is for building/modifying the plugin locally.

### Details
** Bug
    * [OPENENGSB-2408] - LicenseFormat test fails

** Improvement
    * [OPENENGSB-2419] - Move documentation from framework to plugin
    * [OPENENGSB-2440] - Use CHANGELOG instead of CHANGELOG.md

** Library Upgrade
    * [OPENENGSB-2409] - Upgrade to openengsb-root 26
    * [OPENENGSB-2410] - Upgrade commons-compress to 1.3
    * [OPENENGSB-2411] - Upgrade to commons-io 2.1
    * [OPENENGSB-2412] - Include guava library into the maven plugin
    * [OPENENGSB-2441] - Upgrade to root-27

** New Feature
    * [OPENENGSB-2413] - Add infrastructure to extract annotated sources from files into docbook programlistings
    * [OPENENGSB-2415] - Add annotated parser to include java into the docs
    * [OPENENGSB-2416] - Add annotated parser to include xml into the docs
    * [OPENENGSB-2417] - Add annotated parser to include cfg into the docs
    * [OPENENGSB-2418] - Add annotated parser to include properties into the docs
    * [OPENENGSB-2439] - Include source distribution

** Task
    * [OPENENGSB-2248] - Release openengsb-maven-plugin-2.3.0
    * [OPENENGSB-2420] - Document SourceExtractionFeature from user and contributor perspective
    * [OPENENGSB-2438] - Add License file
    * [OPENENGSB-2457] - Temporary deactivate source parsing


openengsb-maven-plugin-2.2.0 2011-10-31
--------------------------------------------

The second feature release of the openengsb-maven-plugin contains one important new feature: automatic version
resolving for openengsb-archetypes. The maven plugin choses now automatically the latest version of the openengsb
and archetypes available in m2 central to use. In addition we've upgraded openengsb-root (24) and junit (4.10).

### Details
** Improvement
    * [OPENENGSB-2245] - Adapt the maven plugin to archetype changes

** Library Upgrade
    * [OPENENGSB-2246] - Upgrade to openengsb-root 24
    * [OPENENGSB-2247] - Upgrade to junit 4.10

** Task
    * [OPENENGSB-2073] - Release openengsb-maven-plugin-2.2.0


openengsb-maven-plugin-2.1.0 2011-09-21
--------------------------------------------

The runtime requirements fail in some builds and make a hassle of builds. Those problems had been fixed. Besides the
openengsb-root pom had been upgraded to version 21.

### Details
** Improvement
    * [OPENENGSB-2068] - Remove all direct requirements to the runtime environment

** Library Upgrade
    * [OPENENGSB-2070] - Upgrade to openengsb-root-21

** Task
    * [OPENENGSB-2045] - Release openengsb-maven-plugin-2.1.0


openengsb-maven-plugin-2.0.0 2011-09-20
--------------------------------------------

This release marks an important milestone in the development of the openengsb-maven-plugin switching to semantical
versioning. Mainly we've fixed three bugs in the openengsb goals. In addition this release enhanced the checkstyle
plugin usage a little bit. Finally we've upgraded various references such as openengsb-root (20), commons-compress (1.2)
and junit (4.9).

### Details
** Bug
    * [OPENENGSB-1850] - in windows checkstyle expects CRLF at end of file
    * [OPENENGSB-1953] - openengsb:provision points to wrong script
    * [OPENENGSB-1990] - openengsb:eclipse does still not work

** Improvement
    * [OPENENGSB-2039] - Instead of strange release versions checkstyle config should be done in root
    * [OPENENGSB-2042] - Remove UncommentedMain check from checkstyle configuration

** Library Upgrade
    * [OPENENGSB-1888] - Upgrade openengsb-root to 19
    * [OPENENGSB-1890] - Upgrade commons-compress to 1.2
    * [OPENENGSB-2037] - Upgrade to openengsb-root-20
    * [OPENENGSB-2040] - Upgrade junit to 4.9
    * [OPENENGSB-2041] - Upgrade pax-runner-platfrom to 1.7.5

** Task
    * [OPENENGSB-1818] - Release openengsb-maven-plugin-2.0.0
    * [OPENENGSB-1889] - Adapt name of openengsb-maven-plugin project name


openengsb-maven-plugin-1.5.0 2011-07-13
--------------------------------------------

Although this release does not contain those killer features, compared to 1.4.x it requires various files to be changed 
in the sources and does not allow a simply switch.

### Details
** Bug
    * [OPENENGSB-1359] - Missing licenses and checkstyle errors
    * [OPENENGSB-1676] - openengsb:eclipse does not work

** Improvement
    * [OPENENGSB-1330] - update connector mojo in openengsb-maven-plugin 
    * [OPENENGSB-1606] - add module for disabling checkstyle via comments to checkstyle.xml
    * [OPENENGSB-1620] - Also handle domainVersion and domainRange in openengsb-maven-plugin

** Library Upgrade
    * [OPENENGSB-1357] - Upgrade to openengsb-root-14
    * [OPENENGSB-1623] - Upgrade to openengsb-root-17
    * [OPENENGSB-1624] - Upgrade pax-runner-platform to 1.7.3
    * [OPENENGSB-1681] - Upgrade maven-license-plugin to 1.9.0
    * [OPENENGSB-1723] - Upgrade pax runner to 1.7.4

** Task
    * [OPENENGSB-1302] - Release openengsb-maven-plugin-1.5.0
    * [OPENENGSB-1650] - Checkstyle rule for newline at end of file

src/main/java/org/openengsb/openengsbplugin/GenConnector.java
/**
 * Licensed to the Austrian Association for Software Tool Integration (AASTI)
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. The AASTI licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.openengsb.openengsbplugin;

import java.util.Arrays;
import java.util.List;
import java.util.Properties;
import java.util.Scanner;

import org.apache.maven.plugin.MojoExecutionException;
import org.openengsb.openengsbplugin.base.MavenExecutorMojo;
import org.openengsb.openengsbplugin.exceptions.NoVersionFoundException;
import org.openengsb.openengsbplugin.tools.MavenExecutor;
import org.openengsb.openengsbplugin.tools.OpenEngSBVersionResolver;
import org.openengsb.openengsbplugin.tools.Tools;

/**
 * guides through the creation of a connector for the OpenEngSB via the connector archetype
 *
 * @goal genConnector
 * @inheritedByDefault false
 * @requiresProject false
 * @aggregator true
 */
public class GenConnector extends MavenExecutorMojo {

    private boolean archetypeCatalogLocalOnly = false;

    // INPUT
    private String domainName;
    private String domainVersion;
    private String domainRange;
    private String domaininterface;
    private String connector;
    private String projectName;
    private String domainGroupId;
    private String domainArtifactId;
    private String scmConnection;
    private String scmDeveloperConnection;
    private String scmUrl;
    private String archetypeVersion;

    // CONSTANTS
    private static final String ARCHETYPE_GROUPID = "org.openengsb.tooling.archetypes";
    private static final String ARCHETYPE_ARTIFACTID = "org.openengsb.tooling.archetypes.connector";

    private static final String DOMAIN_GROUPIDPREFIX = "org.openengsb.domain.";

    private static final String DOMAIN_ARTIFACTIDPREFIX = "openengsb-domain-";

    private static final String CONNECTOR_GROUPID = "org.openengsb.connector";

    private static final String DEFAULT_DOMAIN = "domain";
    private static final String DEFAULT_DOMAIN_VERSION = "1.3.0-SNAPSHOT";
    private static final String DEFAULT_DOMAIN_RANGE = "[1.2,1.3)";
    private static final String SCM_CONNECTION = "scm:git:git://github"
        + ".com/openengsb/openengsb-connector-connectorName.git";
    private static final String SCM_DEVELOPER_CONNECTION =
        "scm:git:git@github.com:openengsb/openengsb-connector-connectorName.git";
    private static final String SCM_URL = "http://github.com/openengsb/openengsb-connector-connectorName";

    @Override
    protected void configure() throws MojoExecutionException {
        OpenEngSBVersionResolver versionResolver = new OpenEngSBVersionResolver();
        try {
            archetypeVersion = versionResolver.getLatestVersion();
        } catch (NoVersionFoundException e) {
            System.err.println("#############################################################");
            System.err.println("AN ERROR OCCURED: " + e.getMessage());
            System.err.println("#############################################################");
            archetypeVersion = "";
        }
        readUserInput();
        MavenExecutor genConnectorExecutor = getNewMavenExecutor(this);
        initializeMavenExecutionProperties(genConnectorExecutor);
        genConnectorExecutor.setRecursive(true);
        addMavenExecutor(genConnectorExecutor);
    }

    @Override
    protected void validateIfExecutionIsAllowed() throws MojoExecutionException {
        throwErrorIfWrapperRequestIsRecursive();
    }

    @Override
    protected void postExec() throws MojoExecutionException {
        System.out.println("DON'T FORGET TO ADD THE CONNECTOR TO YOUR RELEASE/ASSEMBLY PROJECT!");
    }

    private void readUserInput() {
        Scanner sc = new Scanner(System.in);

        System.out.print("Use only local archetypeCatalog? (y/n): ");
        String in = sc.nextLine();
        if (in.equalsIgnoreCase("y")) {
            archetypeCatalogLocalOnly = true;
        }

        archetypeVersion = Tools.readValueFromStdin(sc, "Archetype and OpenEngSB Version", archetypeVersion);
        domainName = Tools.readValueFromStdin(sc, "Domain Name", DEFAULT_DOMAIN);
        domainVersion = Tools.readValueFromStdin(sc, "Domain Version", DEFAULT_DOMAIN_VERSION);
        domainRange = Tools.readValueFromStdin(sc, "Domain Version Range", DEFAULT_DOMAIN_RANGE);
        domaininterface = Tools.readValueFromStdin(sc, "Domain Interface",
            String.format("%s%s", Tools.capitalizeFirst(domainName), "Domain"));
        connector = Tools.readValueFromStdin(sc, "Connector Name", "myconnector");

        projectName = Tools.readValueFromStdin(sc, "Project Name",
            String.format("%s", Tools.capitalizeFirst(connector)));
        domainGroupId = String.format("%s%s", DOMAIN_GROUPIDPREFIX, domainName);
        domainArtifactId = String.format("%s%s", DOMAIN_ARTIFACTIDPREFIX, domainName);
        scmConnection = Tools.readValueFromStdin(sc, "SCM Connection", SCM_CONNECTION);
        scmDeveloperConnection = Tools.readValueFromStdin(sc, "SCM Developer Connection", SCM_DEVELOPER_CONNECTION);
        scmUrl = Tools.readValueFromStdin(sc, "SCM Url", SCM_URL);
    }

    private void initializeMavenExecutionProperties(MavenExecutor executor) {
        List<String> goals = Arrays.asList("archetype:generate");

        Properties userProperties = new Properties();

        userProperties.put("archetypeGroupId", ARCHETYPE_GROUPID);
        userProperties.put("archetypeArtifactId", ARCHETYPE_ARTIFACTID);
        userProperties.put("archetypeVersion", archetypeVersion);

        userProperties.put("domainArtifactId", domainArtifactId);
        userProperties.put("groupId", CONNECTOR_GROUPID);
        userProperties.put("artifactId", String.format("%s.%s", CONNECTOR_GROUPID, connector.toLowerCase()));
        userProperties.put("connectorNameL", connector.toLowerCase());
        userProperties.put("groupId", CONNECTOR_GROUPID);
        userProperties.put("version", archetypeVersion);
        userProperties.put("domainInterface", domaininterface);
        userProperties.put("domain", String.format("%s%s", domainName, "Domain"));
        userProperties.put("domainL", domainName.toLowerCase());
        userProperties.put("domainVersion", domainVersion);
        userProperties.put("domainRange", domainRange);
        userProperties.put("domainName", domainName);
        userProperties.put("package", String.format("%s.%s", CONNECTOR_GROUPID, connector.toLowerCase()));
        userProperties.put("domainPackage", domainGroupId);
        userProperties.put("name", projectName);
        userProperties.put("connectorName", Tools.capitalizeFirst(connector));
        userProperties.put("scmConnection", scmConnection);
        userProperties.put("scmDeveloperConnection", scmDeveloperConnection);
        userProperties.put("scmUrl", scmUrl);

        // local archetype catalog only
        if (archetypeCatalogLocalOnly) {
            userProperties.put("archetypeCatalog", "local");
        }

        executor.addGoals(goals);
        executor.addUserProperties(userProperties);
    }

}
